docs/execution.md

```diff
@@ -0,0 +1,208 @@
+* ExecuteSC spends `max_gas * gas_price + coins` from the sender address in the CSS ledger, which must therefore be available
+* all ExecuteSC operations included in the block must be ordered by decreasing `gas_price`
+* the sum of the max_gas values of all ExecuteSC operations in the block must not be higher than a protocol constant `max_block_gas`
+
+When included in a block, from the CSS point of view, an ExecuteSC operation will only spend `max_gas * gas_price + coins` coins from the sender address and do nothing else.
+
+When observing an ExecuteSC operation, the SCE will:
+* if any of the following steps fail, the bytecode is not executed and all its changes to the SCE ledger are rollbacked. The sender loses `max_gas * gas_price + coins`.
+* parse `bytecode` as smart contract bytecode
+* execute the bytecode in `bytecode`:
+  * if hard config-defined engine resource limits are hit (e.g. too much RAM used), execution fails
+  * if the execution gas usage exceeds `max_gas` (defined inside the ExecuteSC operation call), execution fails
+
+### SCE/CSS pipeline
+
+The SCE and CSS modules are separate.
+
+There is a FIFO message pipe from the CSS to the SCE, and another from the SCE to the CSS. 
+
+```mermaid
+sequenceDiagram
+    CSS->>SCE: BlockcliqueChanged
+    SCE-->>CSS: TransferToConsensus
+```
+
+**Whenever the blockclique changes in any way**, the CSS notifies the SCE of all the blocks of the current blockclique using the `BlockCliqueChanged(BlockHashMap<Block>)` message.
+Note that this behevior is "theoretical" and will be optimized to avoid sending/recomputing everything at every blockclique change.
+Possible optimizations include caching, checkpointing, incremental updates, etc... but are beyond the scope of this document.
+
+**Whenever some blocks become final**, the CSS notifies the SCE of all new final blocks of using the `BlocksFinal(BlockHashMap<Block>)` message.
+
+```rust
+BlockcliqueChanged {
+    blockclique: BlockHashMap<Block>,
+    finalized: BlockHashMap<Block>
+}
+```
+
+The SCE locally maintains a Final SCE ledger and a Candidate SCE ledger.
+Addresses in the CSS ledger are also part of the SCE ledgers.
+Smart-contract specific addresses are also part of the SCE ledger (but not CSS ledgers).
+
+When an SCE block execution requests a coin transfer form the SCE ledger to the CSS ledger, a `TransferToConsensus{target_addr: Address, origin_slot: Slot, amount: Amount}` message is sent from the SCE to the CSS.
+`origin_slot` is the slot of the block causing the `TransferToConsensus`.
+
+When the SCE receives a `BlockCliqueChanged` message:
+* the SCE reset its Candidate SCE ledger to the Final SCE ledger
+* the SCE sorts all blockclique blocks by increasing slot
+* for each block B in that order, the SCE processes the ExecuteSC operations in the block in the order they appear in the block. For each such operation `op`:
+  * credit the block creator in the SCE ledger with `op.max_gas * op.gas_price`
```

Comment:
Why do we credit the block creator in the sce ledger ? And as it's considered as fee why only credit the block creator and not the other fee recipients ?

Reply:
see one of the previous comments. We could consider it a fee, OR we could credit it only on the SCE side... both are options, any preference ?

Reply:
OK let's credit in SCE :)